python/lsst/daf/butler/_butler.py

```diff
@@ -1606,6 +1606,7 @@ def datasetExists(
         ref = self._findDatasetRef(datasetRefOrType, dataId, collections=collections, **kwargs)
         return self.datastore.exists(ref)
 
+    @transactional
```

Comment:
Instead of putting this transaction up here, I think we want to start add a `with self.datastore.transaction()` block just before the registry transaction block, so the `emptyTrash` call happens with it but still after the registry transaction closes.  That will keep us from doing the (slow) querying inside the transaction, and it'll make sure we actually try to commit the registry transaction before starting to empty the trash, which might be necessary to detect some errors there.  I think with the outer `@transactional` the inner registry transaction might not do anything, so we might not see errors until after the trash has been emptied.

Reply:
`emptyTrash` is traditionally not part of the transaction -- we designed it to try to run regardless and it only deletes the rows if the file deletes happen and we can't rollback the file deletes. I'll put the datastore transaction around the registry transaction.